src/Rector/Class_/RenameClassRector.php

```diff
@@ -176,4 +203,96 @@ private function isValidClassNameChange(Node $node, string $newName, Class_ $cla
         }
         return ! (in_array($node, $classNode->implements, true) && class_exists($newName));
     }
+
+    private function refactorNamespaceNode(Namespace_ $node): ?Node
+    {
+        $name = $this->getName($node);
+        if ($name === null) {
+            return null;
+        }
+
+        $classNode = $this->getClassOfNamespaceToRefactor($node);
+        if (! $classNode) {
+            return null;
+        }
+
+        $newClassFqn = $this->oldToNewClasses[$this->getName($classNode)];
+        $newNamespace = $this->getNamespacePart($newClassFqn);
+
+        // Renaming to class without namespace (example MyNamespace\DateTime -> DateTimeImmutable)
+        if (! $newNamespace) {
+            $classNode->name = $newClassFqn;
+
+            return $classNode;
+        }
+
+        $node->name = new Name($newNamespace);
+
+        return $node;
+    }
+
+    private function getClassOfNamespaceToRefactor(Namespace_ $namespace): ?ClassLike
+    {
+        $foundClass = $this->betterNodeFinder->findFirst($namespace, function (Node $node) {
+            if (! $node instanceof ClassLike) {
+                return false;
+            }
+
+            return isset($this->oldToNewClasses[$this->getName($node)]);
+        });
+
+        if ($foundClass instanceof ClassLike) {
+            return $foundClass;
+        }
+
+        return null;
+    }
+
+    private function refactorClassLikeNode(ClassLike $classLike): ?Node
+    {
+        $name = $this->getName($classLike);
+        if ($name === null) {
+            return null;
+        }
+
+        $newName = $this->oldToNewClasses[$name] ?? null;
+        if (! $newName) {
+            return null;
+        }
+
+        if (in_array($name, $this->alreadyProcessed, true)) {
+            return null;
+        }
+
+        $this->alreadyProcessed[] = $name;
+
+        $newName = $this->oldToNewClasses[$name];
+        $newClassNamePart = $this->getClassNamePart($newName);
+        $newNamespacePart = $this->getNamespacePart($newName);
+
+        $classLike->name = new Name($newClassNamePart);
+
+        // Old class did not have any namespace, we need to wrap class with Namespace_ node
+        if ($newNamespacePart && ! $this->getNamespacePart($name)) {
+            return new Namespace_(new Name($newNamespacePart), [$classLike]);
+        }
+
+        return $classLike;
+    }
+
+    private function getClassNamePart(string $fqn): string
+    {
+        $fqn = trim($fqn, '\\');
+        $backslashCount = substr_count($fqn, '\\');
+
+        return $backslashCount === 0 ? $fqn : Strings::after($fqn, '\\', $backslashCount);
+    }
+
+    private function getNamespacePart(string $fqn): ?string
+    {
+        $fqn = trim($fqn, '\\');
+        $backslashCount = substr_count($fqn, '\\');
+
+        return $backslashCount === 0 ? null : Strings::before($fqn, '\\', $backslashCount);
+    }
```

Comment:
Move these 2 last methods to `Rector\CodingStyle\Naming\ClassNaming` and use it here

Reply:
I see, but i would better place it under `Rector\Utils` namespace, what do you think?

Reply:
Ok, i see that class already exists and there is already method doing the same (not covering the edge case, so i will improve it), though i think this could be static function as it is exactly what i think util class would looks like 😄